src/dune_engine/fs_memo.ml

```diff
@@ -280,15 +280,20 @@ let dir_contents ?(force_update = false) path =
    because the result's type depends on [f]. There are only two call sites of
    [with_lexbuf_from_file], so perhaps we could just replace this more general
    function with two simpler ones that can be cached independently. *)
-let with_lexbuf_from_file path ~f =
+let make_reader path ~f =
   let+ () = Watcher.watch ~try_to_watch_via_parent:true path in
   (* This is a bit of a hack. By reading [file_digest], we cause the [path] to
      be recorded in the [Fs_cache.Untracked.file_digest], so the build will be
      restarted if the digest changes. *)
   let (_ : Cached_digest.Digest_result.t) =
     Fs_cache.read Fs_cache.Untracked.file_digest path
   in
-  Io.Untracked.with_lexbuf_from_file path ~f
+  f path
+
+let with_lexbuf_from_file path ~f =
+  make_reader path ~f:(fun path -> Io.Untracked.with_lexbuf_from_file path ~f)
+
+let read_file path = make_reader path ~f:Io.read_file
```

Comment:
To be consistent with naming in the rest of the module, I suggest calling this function `file_contents`.